mono/utils/mono-threads-posix.c

```diff
@@ -161,7 +161,24 @@ mono_threads_pthread_kill (MonoThreadInfo *info, int signum)
 	g_error ("pthread_kill () is not supported by this platform");
 #endif
 
-	if (result && result != ESRCH)
+	/*
+	 * ESRCH just means the thread is gone; this is usually not fatal.
+	 *
+	 * ENOTSUP can occur if we try to send signals (e.g. for sampling) to Grand
+	 * Central Dispatch threads on Apple platforms. This is kinda bad, but
+	 * since there's really nothing we can do about it, we just ignore it and
+	 * move on.
+	 *
+	 * All other error codes are ill-documented and usually stem from various
+	 * OS-specific idiosyncracies. We want to know about these, so fail loudly.
```

Comment:
Signal queue overflow -- does that imply, I know it sounds bad, but backoff and retry?

Reply:
ESRCH can also mean the thread id got corrupted somewhere?  I guess either and impossible to know.

Reply:
> Signal queue overflow -- does that imply, I know it sounds bad, but backoff and retry?

No, it almost always indicates a bug. Signal queue overflow is only a thing in the context of real time signals, i.e. where signal delivery is guaranteed. For normal signals, the OS will just drop signals when the queue overflows. So, if you experience signal queue overflow, it means you're bombarding a thread with signals but it's currently in a state where it cannot handle them.

See ccd373f42986ac5346b305b02beb153c7f311234 for an explanation of the particular problem we ran into in the past.

> ESRCH can also mean the thread id got corrupted somewhere? I guess either and impossible to know.

Right, there is no way to tell the difference.